docs/samples/Microsoft.ML.Samples/Dynamic/MatrixFactorization.cs

```diff
@@ -0,0 +1,102 @@
+        // could be 1.
+        const int _synthesizedMatrixFirstColumnIndex = 1;
+        const int _synthesizedMatrixFirstRowIndex = 1;
+        const int _synthesizedMatrixColumnCount = 60;
+        const int _synthesizedMatrixRowCount = 100;
+
+        // A data structure used to encode a single value in matrix
+        internal class MatrixElement
+        {
+            // Matrix column index starts from _synthesizedMatrixFirstColumnIndex and is at most
+            // _synthesizedMatrixFirstColumnIndex + _synthesizedMatrixColumnCount - 1.
+            // Contieuous=true means that all values between the min and max indexes are all allowed.
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixColumnCount, Min = _synthesizedMatrixFirstColumnIndex)]
+            public uint MatrixColumnIndex;
+            // Matrix row index starts from _synthesizedMatrixFirstRowIndex and is at most
+            // _synthesizedMatrixFirstRowIndex + _synthesizedMatrixRowCount - 1.
+            // Contieuous=true means that all values between the min and max indexes are all allowed.
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixRowCount, Min = _synthesizedMatrixFirstRowIndex)]
+            public uint MatrixRowIndex;
+            // The value at the column MatrixColumnIndex and row MatrixRowIndexin.
+            public float Value;
+        }
+
+        // A data structure used to encode prediction result. Comparing with MatrixElement, The field Value in MatrixElement is
+        // renamed to Score because Score is the default name of matrix factorization's output.
+        internal class MatrixElementForScore
+        {
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixColumnCount, Min = _synthesizedMatrixFirstColumnIndex)]
+            public uint MatrixColumnIndex;
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixRowCount, Min = _synthesizedMatrixFirstRowIndex)]
+            public uint MatrixRowIndex;
+            public float Score;
+        }
+
+        // This example first creates in-memory data and then use it to train a matrix factorization model. Afterward, quality metrics are reported.
+        public static void MatrixFactorizationInMemoryData()
+        {
+            // Create an in-memory matrix as a list of tuples (column index, row index, value).
+            var dataMatrix = new List<MatrixElement>();
+            for (uint i = _synthesizedMatrixFirstColumnIndex; i < _synthesizedMatrixFirstColumnIndex + _synthesizedMatrixColumnCount; ++i)
+                for (uint j = _synthesizedMatrixFirstRowIndex; j < _synthesizedMatrixFirstRowIndex + _synthesizedMatrixRowCount; ++j)
+                    dataMatrix.Add(new MatrixElement() { MatrixColumnIndex = i, MatrixRowIndex = j, Value = (i + j) % 5 });
+
+            // Creating the ML.Net IHostEnvironment object, needed for the pipeline
```

Comment:
>// Creating the ML.Net IHostEnvironment object, needed for the pipeline [](start = 12, length = 71)

// Create a new context for ML.NET operations. It can be used for exception tracking and logging, 
// as a catalog of available operations and as the source of randomness.

Reply:
Changed.

---
In reply to: [229809280](https://github.com/dotnet/machinelearning/pull/1450#discussion_r229809280) [](ancestors = 229809280)